### PR TITLE
http (server,client) OPTIONS method added

### PR DESCRIPTION
**Pull request by sillent**
Hello, 

* feat(http_server/http_client): OPTIONS http method support
* description: for #21169  add support OPTIONS method for http (client, server) sources

### CODE REVIEW

File: website/cue/reference/components/sources/base/http.cue [hunk 1]
```diff
@@ -477,12 +477,13 @@ base: components: sources: http: configuration: {
...
-				HEAD:   "HTTP HEAD method."
-				PATCH:  "HTTP PATCH method."
-				POST:   "HTTP POST method."
-				PUT:    "HTTP Put method."
+				DELETE:  "HTTP DELETE method."
+				GET:     "HTTP GET method."
+				HEAD:    "HTTP HEAD method."
+				PATCH:   "HTTP PATCH method."
+				POST:    "HTTP POST method."
+				PUT:     "HTTP Put method."
```
Comment (buraizu) [inline website/cue/reference/components/sources/base/http.cue:485]:
```suggestion
				PUT:     "HTTP PUT method."
```

File: website/cue/reference/components/sources/base/http_client.cue [hunk 2]
```diff
@@ -429,12 +429,13 @@ base: components: sources: http_client: configuration: {
...
-				HEAD:   "HTTP HEAD method."
-				PATCH:  "HTTP PATCH method."
-				POST:   "HTTP POST method."
-				PUT:    "HTTP Put method."
+				DELETE:  "HTTP DELETE method."
+				GET:     "HTTP GET method."
+				HEAD:    "HTTP HEAD method."
+				PATCH:   "HTTP PATCH method."
+				POST:    "HTTP POST method."
+				PUT:     "HTTP Put method."
```
Comment (buraizu) [inline website/cue/reference/components/sources/base/http_client.cue:437]:
```suggestion
				PUT:     "HTTP PUT method."
```

File: website/cue/reference/components/sources/base/http_server.cue [hunk 3]
```diff
@@ -477,12 +477,13 @@ base: components: sources: http_server: configuration: {
...
-				HEAD:   "HTTP HEAD method."
-				PATCH:  "HTTP PATCH method."
-				POST:   "HTTP POST method."
-				PUT:    "HTTP Put method."
+				DELETE:  "HTTP DELETE method."
+				GET:     "HTTP GET method."
+				HEAD:    "HTTP HEAD method."
+				PATCH:   "HTTP PATCH method."
+				POST:    "HTTP POST method."
+				PUT:     "HTTP Put method."
```
Comment (buraizu) [inline website/cue/reference/components/sources/base/http_server.cue:485]:
```suggestion
				PUT:     "HTTP PUT method."
```

**Review by buraizu — APPROVED** [review]
Approving with some minor suggestions for consistency

**Comment by jszwedko** [comment]
@sillent do you expect the incoming `OPTIONS` requests to have request bodies? I was testing out this PR and noticed that it doesn't emit any events unless the request has a body.

**Comment by sillent** [comment]
> @sillent do you expect the incoming `OPTIONS` requests to have request bodies? I was testing out this PR and noticed that it doesn't emit any events unless the request has a body.

Requests with body and without works fine - events triggered

request:
`curl localhost:8080 -X OPTIONS`
event:
`{"accept":"*/*","host":"localhost:8080","message":"","path":"/","source_type":"http_server","timestamp":"2024-09-
07T14:09:13.203418417Z","user-agent":"curl/7.81.0"}`

request with body: 
`curl localhost:8080 -X OPTIONS -d '{"json": true, "toml": false}'`
event:
`{"accept":"*/*","content-length":"29","content-type":"application/x-www-form-urlencoded","host":"localhost:8080",
"message":"{\"json\": true, \"toml\": false}","path":"/","source_type":"http_server","timestamp":"2024-09-07T14:0
9:19.793557009Z","user-agent":"curl/7.81.0"}`

**Comment by jszwedko** [comment]
> > @sillent do you expect the incoming `OPTIONS` requests to have request bodies? I was testing out this PR and noticed that it doesn't emit any events unless the request has a body.
> 
> Requests with body and without works fine - events triggered
> 
> request: `curl localhost:8080 -X OPTIONS` event: `{"accept":"*/*","host":"localhost:8080","message":"","path":"/","source_type":"http_server","timestamp":"2024-09- 07T14:09:13.203418417Z","user-agent":"curl/7.81.0"}`
> 
> request with body: `curl localhost:8080 -X OPTIONS -d '{"json": true, "toml": false}'` event: `{"accept":"*/*","content-length":"29","content-type":"application/x-www-form-urlencoded","host":"localhost:8080", "message":"{\"json\": true, \"toml\": false}","path":"/","source_type":"http_server","timestamp":"2024-09-07T14:0 9:19.793557009Z","user-agent":"curl/7.81.0"}`

Hmm, I still don't see any messages published when I try to run the code in this PR with this config:

```yaml
sources:
  source0:
    type: http_server
    address: 0.0.0.0:8080
    method: OPTIONS
sinks:
  sink0:
    inputs:
    - source0
    type: console
    encoding:
      codec: json
```

and send a request without a body:

```
❯ curl localhost:8080 -X OPTIONS -v
* Host localhost:8080 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8080...
* connect to ::1 port 8080 from ::1 port 57825 failed: Connection refused
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080
> OPTIONS / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/8.7.1
> Accept: */*
>
* Request completely sent off
< HTTP/1.1 200 OK
< content-length: 0
< date: Fri, 13 Sep 2024 21:04:13 GMT
<
* Connection #0 to host localhost left intact
```

And in Vector, nothing:

```
❯ ./target/debug/vector --config /tmp/tmp.yaml
2024-09-13T21:05:34.111173Z  INFO vector::app: Log level is enabled. level="info"
2024-09-13T21:05:34.112297Z  INFO vector::app: Loading configs. paths=["/tmp/tmp.yaml"]
2024-09-13T21:05:34.116907Z  INFO vector::topology::running: Running healthchecks.
2024-09-13T21:05:34.117025Z  INFO vector::topology::builder: Healthcheck passed.
2024-09-13T21:05:34.117254Z  INFO vector: Vector has started. debug="true" version="0.41.0" arch="aarch64" revision=""
2024-09-13T21:05:34.117249Z  INFO source{component_kind="source" component_id=source0 component_type=http_server}: vector::sources::util::http::prelude: Building HTTP server. address=0.0.0.0:8080
```

Only when I send a request body do I get messages output.

**Comment by sillent** [comment]
@jszwedko 

I reproduced it by my self once again, and find out - 

If in your config example replace OPTION with something other, like POST, - nothing changes, events for requests without body also should not be presented

But in my configuration, i use `encoding: binary` in sources, and events appears for requests with and without body for OPTIONS and other  methods

p.s. my example configuration

```toml
[sources.http_options]
type = "http_server"
address = "0.0.0.0:8080"
encoding = "binary"
method = "OPTIONS"

[sinks.mysinkfile]
type = "console"
inputs = ["http_options"]
buffer.type = "memory"
encoding.codec = "json"
```

**Comment by jszwedko** [comment]
> @jszwedko
> 
> I reproduced it by my self once again, and find out -
> 
> If in your config example replace OPTIONS with something other, like POST, - nothing changes, events for requests without body also should not be presented
> 
> But in my configuration, i use `encoding: binary` in sources, and events appears for requests with and without body for OPTIONS and other methods
> 
> p.s. my example configuration
> 
> ```toml
> [sources.http_options]
> type = "http_server"
> address = "0.0.0.0:8080"
> encoding = "binary"
> method = "OPTIONS"
> 
> [sinks.mysinkfile]
> type = "console"
> inputs = ["http_options"]
> buffer.type = "memory"
> encoding.codec = "json"
> ```

Ah, gotcha, I see. That makes sense: the `binary` encoding just takes whatever the raw payload is.

**Review by jszwedko — APPROVED** [review]
Thanks for this contribution!

**Review by jszwedko — COMMENTED** [review]
Could you add a changelog entry per https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md?

**Comment by sillent** [comment]
> Could you add a changelog entry per https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md?

done

**Comment by jszwedko** [comment]
Thanks!

**Comment by sillent** [comment]
fix component docs for ci

**Pull request merged**